File: lib/services/anomalies.dart
import 'dart:math' as math;

import '../models/category.dart';
import '../models/notice.dart';
import '../models/transaction.dart';
import '../ui/theme.dart';
import 'dates.dart';

class SampleStats {
  const SampleStats({
    required this.n,
    required this.mean,
    required this.median,
    required this.stdev,
  });

  final int n;
  final double mean;
  final double median;
  final double stdev;
}

class _Spike {
  const _Spike({required this.severity, required this.multiple});
  final AnomalySeverity severity;
  final double multiple;
}

class AnomalyEngine {
  const AnomalyEngine();

  SampleStats sampleStats(List<double> values) {
    final n = values.length;
    if (n == 0) {
      return const SampleStats(n: 0, mean: 0, median: 0, stdev: 0);
    }
    final sorted = [...values]..sort();
    final mean = values.reduce((a, b) => a + b) / n;
    final mid = sorted.length ~/ 2;
    final median = sorted.length.isEven
        ? (sorted[mid - 1] + sorted[mid]) / 2
        : sorted[mid];
    final variance = n < 2
        ? 0.0
        : values.fold<double>(0, (sum, value) => sum + (value - mean) * (value - mean)) /
            (n - 1);
    return SampleStats(
      n: n,
      mean: mean,
      median: median,
      stdev: variance <= 0 ? 0 : math.sqrt(variance),
    );
  }

  String asOfDate(String viewMonth, DateTime today) {
    return monthKey(today) == viewMonth ? dateKey(today) : endOfMonth(viewMonth);
  }

  List<String> priorMonths(String viewMonth, {int count = 8}) {
    return [for (var i = count; i >= 1; i--) addMonths(viewMonth, -i)];
  }

  bool isStableBill(SampleStats stats) {
    return stats.n >= 3 && stats.mean > 0 && stats.stdev / stats.mean < 0.12;
  }

  _Spike? _classifySpike(double current, SampleStats baseline) {
    if (baseline.n < 3 || current < 30) return null;
    final multiple = baseline.median > 0 ? current / baseline.median : 0.0;
    final z = baseline.stdev > 1 ? (current - baseline.mean) / baseline.stdev : 0.0;
    final delta = current - baseline.median;
    if (delta < 40 || multiple < 1.25) return null;
    if (isStableBill(baseline) && multiple < 1.5) return null;
    if (multiple >= 1.75 || (z >= 2.15 && multiple >= 1.5)) {
      return _Spike(severity: AnomalySeverity.unusual, multiple: multiple);
    }
    if (multiple >= 1.35 || z >= 1.4) {
      return _Spike(severity: AnomalySeverity.watch, multiple: multiple);
    }
    return null;
  }

  _Spike? _classifyCharge(double amount, SampleStats baseline) {
    if (baseline.n < 4 || amount < 40) return null;
    final multiple = baseline.median > 0 ? amount / baseline.median : 0.0;
    final z = baseline.stdev > 1 ? (amount - baseline.mean) / baseline.stdev : 0.0;
    if (amount - baseline.median < 25) return null;
    if (multiple < 1.8) return null;
    if (multiple >= 2.5 || z >= 2.4) {
      return _Spike(severity: AnomalySeverity.unusual, multiple: multiple);
    }
    if (multiple >= 2.1 || z >= 2) {
      return _Spike(severity: AnomalySeverity.watch, multiple: multiple);
    }
    return null;
  }

  double spentThrough(
    List<MoneyTx> txs,
    String month,
    String throughDate, {
    String? categoryId,
  }) {
    var total = 0.0;
    for (final tx in txs) {
      if (!tx.isExpense) continue;
      if (monthKey(tx.bookedAt) != month) continue;
      if (dateKey(tx.bookedAt).compareTo(throughDate) > 0) continue;
      if (categoryId != null && tx.categoryId != categoryId) continue;
      total += tx.absAmount;
    }
    return roundMoney(total);
  }

  List<SpendingAnomaly> findSpendingAnomalies({
    required List<MoneyTx> txs,
    required String viewMonth,
    required DateTime today,
  }) {
    final through = asOfDate(viewMonth, today);
    final history = priorMonths(viewMonth);
    final pointInMonth = monthKey(today) == viewMonth;
    final anomalies = <SpendingAnomaly>[];

    final overallHistory = history
        .map((month) => spentThrough(txs, month, endOfMonth(month)))
        .where((value) => value > 0)
        .toList();
    final overallNow = spentThrough(txs, viewMonth, through);
    final overallStats = sampleStats(overallHistory);
    final overallSpike = _classifySpike(overallNow, overallStats);
    if (overallSpike != null) {
      anomalies.add(
        SpendingAnomaly(
          id: 'month-spike:$viewMonth',
          kind: AnomalyKind.monthSpike,
          severity: overallSpike.severity,
          categoryId: 'overall',
          label: 'Visos išlaidos',
          month: viewMonth,
          amount: overallNow,
          baseline: overallStats.median,
          multiple: overallSpike.multiple,
          message: _monthSpikeMessage(
            overallNow,
            overallStats.median,
            overallSpike.multiple,
            pointInMonth,
            viewMonth,
          ),
        ),
      );
    }

    for (final category in Categories.spendable) {
      final prior = history
          .map(
            (month) => spentThrough(
              txs,
              month,
              endOfMonth(month),
              categoryId: category.id,
            ),
          )
          .where((value) => value > 0)
          .toList();
      final current = spentThrough(
        txs,
        viewMonth,
        through,
        categoryId: category.id,
      );
      final stats = sampleStats(prior);
      final spike = _classifySpike(current, stats);
      if (spike != null) {
        anomalies.add(
          SpendingAnomaly(
            id: 'category-spike:${category.id}:$viewMonth',
            kind: AnomalyKind.categorySpike,
            severity: spike.severity,
            categoryId: category.id,
            label: category.name,
            month: viewMonth,
            amount: current,
            baseline: stats.median,
            multiple: spike.multiple,
            message: _categorySpikeMessage(
              category.name,
              current,
              stats.median,
              spike.multiple,
              pointInMonth,
            ),
          ),
        );
      }

      if (category.id == 'housing' || category.id == 'utilities') {
        continue;
      }

      final historyCharges = txs
          .where(
            (tx) =>
                tx.isExpense &&
                tx.categoryId == category.id &&
                monthKey(tx.bookedAt).compareTo(viewMonth) < 0,
          )
          .map((tx) => tx.absAmount)
          .toList();
      final chargeStats = sampleStats(historyCharges);
      final monthCharges = txs.where(
        (tx) =>
            tx.isExpense &&
            tx.categoryId == category.id &&
            monthKey(tx.bookedAt) == viewMonth &&
            dateKey(tx.bookedAt).compareTo(through) <= 0,
      );
      for (final tx in monthCharges) {
        final charge = _classifyCharge(tx.absAmount, chargeStats);
        if (charge == null) continue;
        anomalies.add(
          SpendingAnomaly(
            id: 'large-charge:${tx.id}',
            kind: AnomalyKind.largeCharge,
            severity: charge.severity,
            categoryId: category.id,
            label: category.name,
            month: viewMonth,
            amount: tx.absAmount,
            baseline: chargeStats.median,
            multiple: charge.multiple,
            message: _chargeMessage(
              category.name,
              tx,
              chargeStats.median,
              charge.multiple,
            ),
            transactionId: tx.id,
          ),
        );
      }
    }

    anomalies.sort((a, b) {
      final rank = _rank(b.severity) - _rank(a.severity);
      if (rank != 0) return rank;
      return b.amount.compareTo(a.amount);
    });
    return anomalies;
  }

  String _categorySpikeMessage(
    String name,
    double current,
    double baseline,
    double multiple,
    bool pointInMonth,
  ) {
    final timing = pointInMonth
        ? 'iki šios mėnesio dienos'
        : 'per visą mėnesį';
    return '$name yra ${multiple.toStringAsFixed(1)}× įprastos ${formatEur(baseline)} $timing (${formatEur(current)} vs ${formatEur(baseline)}).';
  }

  String _monthSpikeMessage(
    double current,
    double baseline,
    double multiple,
    bool pointInMonth,
    String viewMonth,
  ) {
    final extra = current - baseline;
    if (pointInMonth) {
      return 'Bendros išlaidos jau ${formatEur(extra)} viršija įprastą mėnesį (${formatEur(current)} vs ${formatEur(baseline)}).';
    }
    return '${_monthLabel(viewMonth)} išlaidos yra ${multiple.toStringAsFixed(1)}× įprastos ${formatEur(baseline)}.';
  }

  String _chargeMessage(
    String name,
    MoneyTx tx,
    double baseline,
    double multiple,
  ) {
    final note = tx.description.isNotEmpty &&
            tx.description.toLowerCase() != tx.merchant.toLowerCase()
        ? ' „${tx.description}“'
        : '';
    return '$name$note turi ${formatEur(tx.absAmount)} mokėjimą — apie ${multiple.toStringAsFixed(1)}× įprastos ${formatEur(baseline)} eilutės.';
  }

  String _monthLabel(String month) {
    final parts = month.split('-');
    return '${parts[1]}.${parts[0]}';
  }

  int _rank(AnomalySeverity severity) =>
      severity == AnomalySeverity.unusual ? 2 : 1;
}

File: src/lib/anomalies.ts
import { spentThrough } from './analytics.ts'
import { addMonths, endOfMonth, monthKey, monthLabel } from './dates.ts'
import { formatMoney } from './money.ts'
import type { AppState, SpendingAnomaly, Transaction } from '../types.ts'

export type SampleStats = {
  n: number
  mean: number
  median: number
  stdev: number
}

export function sampleStats(values: number[]): SampleStats {
  const n = values.length
  if (n === 0) return { n: 0, mean: 0, median: 0, stdev: 0 }
  const sorted = [...values].sort((a, b) => a - b)
  const mean = values.reduce((sum, value) => sum + value, 0) / n
  const mid = Math.floor(sorted.length / 2)
  const median = sorted.length % 2 === 0 ? (sorted[mid - 1] + sorted[mid]) / 2 : sorted[mid]
  const variance =
    n < 2 ? 0 : values.reduce((sum, value) => sum + (value - mean) ** 2, 0) / (n - 1)
  return { n, mean, median, stdev: Math.sqrt(variance) }
}

export function asOfDate(viewMonth: string, today: string): string {
  return today.startsWith(viewMonth) ? today : endOfMonth(viewMonth)
}

function priorMonths(viewMonth: string, count = 8): string[] {
  const months: string[] = []
  for (let i = count; i >= 1; i -= 1) months.push(addMonths(viewMonth, -i))
  return months
}

function isStableBill(stats: SampleStats): boolean {
  return stats.n >= 3 && stats.mean > 0 && stats.stdev / stats.mean < 0.12
}

function classifySpike(current: number, baseline: SampleStats): {
  severity: SpendingAnomaly['severity']
  multiple: number
} | null {
  if (baseline.n < 3 || current < 30) return null
  if (isStableBill(baseline) && current < baseline.mean * 1.18) return null
  const multiple = baseline.median > 0 ? current / baseline.median : 0
  const z = baseline.stdev > 1 ? (current - baseline.mean) / baseline.stdev : 0
  const delta = current - baseline.median
  if (delta < 35) return null
  if (multiple >= 1.9 || z >= 2.15) return { severity: 'unusual', multiple }
  if (multiple >= 1.45 && (z >= 1.15 || multiple >= 1.55)) return { severity: 'watch', multiple }
  return null
}

function classifyCharge(amount: number, baseline: SampleStats): {
  severity: SpendingAnomaly['severity']
  multiple: number
} | null {
  if (baseline.n < 4 || amount < 40) return null
  const multiple = baseline.median > 0 ? amount / baseline.median : 0
  const z = baseline.stdev > 1 ? (amount - baseline.mean) / baseline.stdev : 0
  if (amount - baseline.median < 25) return null
  if (multiple >= 2.5 || z >= 2.4) return { severity: 'unusual', multiple }
  if (multiple >= 2.1 || z >= 2) return { severity: 'watch', multiple }
  return null
}

function monthWindow(viewMonth: string, today: string) {
  const through = asOfDate(viewMonth, today)
  const history = priorMonths(viewMonth)
  const pointInMonth = today.startsWith(viewMonth)
  return { through, history, pointInMonth }
}

export function findSpendingAnomalies(
  state: AppState,
  viewMonth: string,
  today: string,
): SpendingAnomaly[] {
  const { through, history, pointInMonth } = monthWindow(viewMonth, today)
  const anomalies: SpendingAnomaly[] = []
  const expenseCategories = state.categories.filter((category) => category.kind === 'expense')

  const overallHistory = history
    .map((month) => spentThrough(state.transactions, month, endOfMonth(month)))
    .filter((value) => value > 0)
  const overallNow = spentThrough(state.transactions, viewMonth, through)
  const overallStats = sampleStats(overallHistory)
  const overallSpike = classifySpike(overallNow, overallStats)
  if (overallSpike) {
    anomalies.push({
      id: `month-spike:${viewMonth}`,
      kind: 'month-spike',
      severity: overallSpike.severity,
      categoryId: 'overall',
      label: 'Overall spending',
      month: viewMonth,
      amount: overallNow,
      baseline: overallStats.median,
      multiple: overallSpike.multiple,
      message: monthSpikeMessage(overallNow, overallStats.median, overallSpike.multiple, pointInMonth, viewMonth),
    })
  }

  for (const category of expenseCategories) {
    const prior = history
      .map((month) => spentThrough(state.transactions, month, endOfMonth(month), category.id))
      .filter((value) => value > 0)
    const current = spentThrough(state.transactions, viewMonth, through, category.id)
    const stats = sampleStats(prior)
    const spike = classifySpike(current, stats)
    if (spike) {
      anomalies.push({
        id: `category-spike:${category.id}:${viewMonth}`,
        kind: 'category-spike',
        severity: spike.severity,
        categoryId: category.id,
        label: category.name,
        month: viewMonth,
        amount: current,
        baseline: stats.median,
        multiple: spike.multiple,
        message: categorySpikeMessage(category.name, current, stats.median, spike.multiple, pointInMonth),
      })
    }

    const historyCharges = state.transactions
      .filter(
        (transaction) =>
          transaction.kind === 'expense' &&
          transaction.categoryId === category.id &&
          monthKey(transaction.date) < viewMonth,
      )
      .map((transaction) => transaction.amount)
    const chargeStats = sampleStats(historyCharges)
    const monthCharges = state.transactions.filter(
      (transaction) =>
        transaction.kind === 'expense' &&
        transaction.categoryId === category.id &&
        inView(transaction, viewMonth, through),
    )
    for (const transaction of monthCharges) {
      const charge = classifyCharge(transaction.amount, chargeStats)
      if (!charge) continue
      anomalies.push({
        id: `large-charge:${transaction.id}`,
        kind: 'large-charge',
        severity: charge.severity,
        categoryId: category.id,
        label: category.name,
        month: viewMonth,
        amount: transaction.amount,
        baseline: chargeStats.median,
        multiple: charge.multiple,
        message: chargeMessage(category.name, transaction, chargeStats.median, charge.multiple),
        transactionId: transaction.id,
      })
    }
  }

  return anomalies.sort((a, b) => rank(b.severity) - rank(a.severity) || b.amount - a.amount)
}

function inView(transaction: Transaction, viewMonth: string, through: string): boolean {
  return monthKey(transaction.date) === viewMonth && transaction.date <= through
}

function rank(severity: SpendingAnomaly['severity']): number {
  return severity === 'unusual' ? 2 : 1
}

function categorySpikeMessage(
  name: string,
  current: number,
  baseline: number,
  multiple: number,
  pointInMonth: boolean,
): string {
  const timing = pointInMonth ? 'by this point in the month' : 'for a full month'
  return `${name} is ${multiple.toFixed(1)}× the usual ${formatMoney(baseline)} ${timing} (${formatMoney(current)} vs ${formatMoney(baseline)}).`
}

function monthSpikeMessage(
  current: number,
  baseline: number,
  multiple: number,
  pointInMonth: boolean,
  viewMonth: string,
): string {
  const extra = current - baseline
  return pointInMonth
    ? `Total spending is already ${formatMoney(extra)} above a usual month (${formatMoney(current)} vs ${formatMoney(baseline)}).`
    : `${monthLabel(viewMonth)} spending is ${multiple.toFixed(1)}× the usual ${formatMoney(baseline)}.`
}

function chargeMessage(
  name: string,
  transaction: Transaction,
  baseline: number,
  multiple: number,
): string {
  const note = transaction.note ? ` “${transaction.note}”` : ''
  return `${name}${note} has a ${formatMoney(transaction.amount, true)} charge — about ${multiple.toFixed(1)}× a typical ${formatMoney(baseline)} line.`
}
